### PR TITLE
Issue 83458:Changed the return check for GetNodeInfo

### LINKED ISSUE #83458: Scheduler cache.GetNodeInfo should return IsNotFound error if applicable
**What happened**:
This [PR](https://github.com/kubernetes/kubernetes/commit/30b0f8bf3bed2c1c146a85ecdfc6b9cb18c5a303) switches NodeLister to scheduler cache for `GetNodeInfo`.  However, the new `GetNodeInfo` implementation in the cache never returns IsNotFound error any more. Rather, it returns a normal error:
https://github.com/kubernetes/kubernetes/blob/67d750bb2893d7a66d45d435dfc075fb17cc15ee/pkg/scheduler/internal/cache/cache.go#L696-L706

However, we rely on the error type for logging, for example:
https://github.com/kubernetes/kubernetes/blob/7cb4850699a4a3c1008382c8b05f245b66ce19ad/pkg/scheduler/algorithm/predicates/predicates.go#L1386-L1394
**What you expected to happen**:
The cache `GetNodeInfo()` should return IsNotFound error if applicable.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

/sig scheduling
@draveness  cc @ahg-g

### PR DESCRIPTION
**Pull request by phiremande**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR fixes the issue 83458

**Which issue(s) this PR fixes**:

Fixes #83458

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/scheduler/internal/cache/cache.go [hunk 1]
```diff
@@ -699,7 +700,7 @@ func (cache *schedulerCache) GetNodeInfo(nodeName string) (*v1.Node, error) {
 
 	n, ok := cache.nodes[nodeName]
 	if !ok {
-		return nil, fmt.Errorf("node %q not found in cache", nodeName)
+		return nil, errors.NewNotFound(v1.Resource("node"), nodeName)
```
Comment (draveness) [inline pkg/scheduler/internal/cache/cache.go]: We don't use errors.NewNotFound to wrap errors in the scheduler package.

**Review by draveness — CHANGES_REQUESTED** [review]

File: pkg/scheduler/internal/cache/cache.go (hunk 1 above)
Comment (phiremande) [reply pkg/scheduler/internal/cache/cache.go]: I am new and hence don't have the full context.  But this is the earlier code. This  used to return API errors.

 func (c *CachedNodeInfo) GetNodeInfo(id string) (*v1.Node, error) { 
 	node, err := c.Get(id) 
  
 	if apierrors.IsNotFound(err) { 
 		return nil, err 
 	}

Comment (phiremande) [reply pkg/scheduler/internal/cache/cache.go]: @draveness Please let me know if the earlier code was a mistake and newer code doesn't want to return API errors. In that case, the issue raised would become a non-issue.

Comment (liu-cong) [reply pkg/scheduler/internal/cache/cache.go]: This function returns either nil or error so no need to use a specific error type. Instead, in 

https://github.com/kubernetes/kubernetes/blob/7cb4850699a4a3c1008382c8b05f245b66ce19ad/pkg/scheduler/algorithm/predicates/predicates.go#L1386-L1394

We can replace the `if apierrors.IsNotFound(err)` with just `if err != nil`.

@draveness WDYT?

Comment (draveness) [reply pkg/scheduler/internal/cache/cache.go]: > This function returns either nil or error so no need to use a specific error type. Instead, in
> 
> https://github.com/kubernetes/kubernetes/blob/7cb4850699a4a3c1008382c8b05f245b66ce19ad/pkg/scheduler/algorithm/predicates/predicates.go#L1386-L1394
> 
> We can replace the `if apierrors.IsNotFound(err)` with just `if err != nil`.
> 
> @draveness WDYT?

+1

Comment (phiremande) [reply pkg/scheduler/internal/cache/cache.go]: Thanks @liu-cong , @draveness I have raised another PR - https://github.com/kubernetes/kubernetes/pull/84077

Comment (draveness) [reply pkg/scheduler/internal/cache/cache.go]: You could update this one by just force pushing this branch.

Comment (phiremande) [reply pkg/scheduler/internal/cache/cache.go]: @draveness , @liu-cong I have updated the current branch. Please review.

**Review by draveness — COMMENTED** [review]
/ok-to-test

**Review by liu-cong — COMMENTED** [review]
By doing a search, I found another place where the caller expects IsNotFoundErr. Please fix that too.
https://github.com/kubernetes/kubernetes/blob/9769d49bb566454617a3df2ab5de6204d3abdaaa/pkg/scheduler/algorithm/priorities/interpod_affinity.go#L116-L122

**Review by draveness — COMMENTED** [review]
Looks good, please squash

**Comment by phiremande** [comment]
@draveness , updated. Please review.

**Review by draveness — APPROVED** [review]
/priority backlog
/lgtm

Thanks!

**Comment by phiremande** [comment]
@draveness , @liu-cong , the tests have passed. Please provide approval.

**Comment by liu-cong** [comment]
Please update the tile of the PR - we don't use `IsNotFound`  error any more.

You can ask ahg-g to approve.

**Comment by phiremande** [comment]
@ahg-g , kindly provide approval.

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 2]
```diff
@@ -1400,12 +1399,9 @@ func (c *PodAffinityChecker) getMatchingAntiAffinityTopologyPairsOfPods(pod *v1.
...
 		existingPodNode, err := c.info.GetNodeInfo(existingPod.Spec.NodeName)
 		if err != nil {
-			if apierrors.IsNotFound(err) {
-				klog.Errorf("Pod %s has NodeName %q but node is not found",
-					podName(existingPod), existingPod.Spec.NodeName)
-				continue
-			}
-			return nil, err
+			klog.Errorf("Pod %s has NodeName %q but node is not found",
+				podName(existingPod), existingPod.Spec.NodeName)
```
Comment (ahg-g) [inline pkg/scheduler/algorithm/predicates/predicates.go]: nit: can you please place this on the same line above.

**Comment by phiremande** [comment]
@liu-cong , @draveness and @ahg-g , Thanks for your guidance and helping me merge my first PR into kubernetes code base. Looking forward to making further contributions (initially with good first issues  and getting myself familiarized with the code base). Thanks once again.

**Pull request merged**